proposals/p3763.md

```diff
@@ -0,0 +1,755 @@
+
+Specifically, we add the following rule:
+
+-   In a declarative scope, it is an error if a name is first looked up and not
+    found, and later introduced.
+
+_Declarative scopes_ here are scopes like namespaces, classes, and interfaces.
+Another way of looking at this rule is that reordering the declarations in a
+declarative scope should never result in a program that is still valid but has a
+different meaning.
+
+It is also possible to declare an entity in a non-declarative scope, such as in
+a function body. We call non-declarative scopes _sequential scopes_, and use a
+different rule:
+
+-   It is an error to redeclare an entity in a sequential scope.
+
+This rule is provisional: no alternative has been suggested, and we don't know
+whether there is strong motivation for supporting redeclaration in sequential
+scopes. In C++, such redeclarations are permitted, but we are not aware of any
+instance of this functionality being used. There are likely other good options
+to use here, but we consider them outside the scope of this proposal.
+
+This rule applies only within a single file. For an entity that is declared in
+an `api` file and redeclared in an `impl` file, an additional rule is imposed:
+
+-   If an entity is declared as non-`extern` in an `api` file and redeclared in
+    an `impl` file of the same library, it is an error for the declaration to
+    use any unqualified name that was introduced in the `impl` file.
+
+It is, however, valid for such a declaration to use a name that was introduced
+in the `api` file and redeclared in the `impl` file. This ensures that the
+declaration has the same unqualified lookup results as a declaration in the
+`api` file.
+
+[Additional rules](#extern-declarations) apply to `extern` declarations.
+
+#### `impl`s that have not yet been declared
+
+It is possible for a declaration of an entity to have one meaning because `impl`
+lookup finds a broad `impl`, and for a redeclaration to have a second meaning
+because `impl` lookup finds a narrower `impl`. However, in such a case, the
+declaration of the narrower `impl` declaration is invalid, because it would
+change the meaning of an earlier `impl` lookup. From #875:
+
+> When an `impl` needs to be resolved, only those `impl` declarations that have
+> that appear earlier are considered. However, if a later `impl` declaration
+> would change the result of any earlier `impl` lookup, the program is invalid.
+
+### Declaration modifiers
+
+Declaration modifier keywords appear prior to the introducer keyword and scope
+in a declaration, so are not involved in checking whether two declarations
+differ. There may be other rules that determine whether modifiers may or must be
+repeated on a redeclaration, but they are out of scope for this proposal.
+
+```carbon
+class A {
+  virtual fn F[self: Self]();
+}
+
+// OK, `virtual` not required to match.
+fn A.F[self: Self]() {}
+```
+
+### `_` parameter names and `unused` modifier
```

Comment:
I should ask here, do we want to continue to allow `_` parameters in extern declarations?

With `_`, a library could rename a function parameter with per-file changes, even if one of those steps is to change the name in `extern` declarations to `_`. Without `_`, the entire package must change in sync.

If the perspective is that this package-scope refactoring requirement is okay for function parameters, maybe we should re-evaluate how other merging works, such as whether redundant redeclarations are allowed? Most of the reason for those had been about avoiding package-scope changes, as I understood it. Is this just a split trade-off, weighed differently here?

Reply:
`_` doesn't allow renaming function parameters in general -- specifically, function parameters that are used later in the signature (eg, type parameters) aren't helped by that. So it's only a half-measure, and a general parameter renaming tool still needs to make package-wide changes. Allowing `_` also makes a purely-syntactic check more complex -- we'd have to disallow later uses of the parameter name somehow, because those might resolve to something else in the context of the declaration using `_`.

I think I wouldn't want to encourage a proliferation of `extern` declarations in arbitrary other parts of a package -- I'd want to encourage a style where there is a single `extern` declaration that's kept alongside the library that owns the real declaration. So to the extent that a change has package-wide impact, that seems like an avoidable problem in the existing code. (Though I do wonder whether we should have some enforcement that people don't create this style issue.)

Reply:
Marking this resolved, I think we have other things to discuss regarding `extern` anyways.